llvm/include/llvm/ADT/StringExtras.h

```diff
@@ -417,8 +418,13 @@ inline std::string join_impl(IteratorT Begin, IteratorT End,
     return S;
 
   size_t Len = (std::distance(Begin, End) - 1) * Separator.size();
-  for (IteratorT I = Begin; I != End; ++I)
-    Len += (*I).size();
+  for (IteratorT I = Begin; I != End; ++I) {
+    if constexpr (std::is_same_v<std::remove_reference_t<decltype(*I)>,
+                                 const char *>)
+      Len += strlen(*I);
+    else
+      Len += (*I).size();
```

Comment:
any chance of using StringRef in the callers? Rather than having more places dealing with null terminated strings?

Reply:
Sure, constructing a temporary StringRef would work too, what's the difference?

> any chance of using StringRef in the callers? Rather than having more places dealing with null terminated strings?

Converting Driver and others to use StringRef seems like a helpful project, though not one I personally have time for. Having a few functions not work with `char*` doesn't seem like a useful state in the meanwhile, but I already wrote the concat-loop where this was actually needed, so I can drop this if we'd rather not have it.

Reply:
> Sure, constructing a temporary StringRef would work too, what's the difference?

It already handles any c-string type without having to branch on type traits. IIUC, the current version does not handle `char *` or char arrays?

Reply:
oh of course, *always* use stringref. thanks!

Reply:
> Sure, constructing a temporary StringRef would work too, what's the difference?

It does allow the implementation to remain generic, if somewhat more subtle than it currently is.

> > any chance of using StringRef in the callers? Rather than having more places dealing with null terminated strings?
> 
> Converting Driver and others to use StringRef seems like a helpful project, though not one I personally have time for. Having a few functions not work with `char*` doesn't seem like a useful state in the meanwhile, but I already wrote the concat-loop where this was actually needed, so I can drop this if we'd rather not have it.

*nod* I'm more comfortable with it with the StringRef/generic implementation. "supports anything that's StringRef-able" sounds OK to me.